### PR TITLE
feat: estimation sub-command for microsatellite instability (msi)  

### PR DESCRIPTION
**Pull request by rohan-ibn-tariq**
##### Overview
Implements microsatellite instability (MSI) detection using an approach that provides subclonal resolution and uncertainty quantification, going beyond traditional binary MSI-H/MSS classification.

##### Core Functionality
**MSI scoring**: Dynamic programming-based posterior probability calculation for MSI status
**Subclonal resolution**: Track MSI evolution across allele frequency thresholds (pseudotime analysis)
**Uncertainty quantification**: Compute credible intervals for MSI scores using exact decimal arithmetic
**Flexible outputs**: TSV data files and Vega-Lite visualization specs for both distribution and pseudotime analyses

##### CLI Command
varlociraptor estimate msi \
  microsatellites.bed \
  vcf tumor.bcf \
  --msi-threshold 3.5 \
  --plot-distribution out1.vl.json \
  --plot-pseudotime out2.vljson \
  --data-distribution out1.tsv \
  --data-pseudotime out2.tsv

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [ ] The documentation at https://github.com/varlociraptor/varlociraptor.github.io is updated in a separate PR to reflect the changes or this is not necessary (e.g. if the change does neither modify the calling grammar nor the behavior or functionalities of Varlociraptor).

## Summary by CodeRabbit

* **New Features**
  * MSI estimation workflow added to the CLI with per-sample AF‑evolution analysis, TSV exports and Vega‑Lite plot specs (distribution & pseudotime).

* **Improvements**
  * Enhanced BED/VCF/BCF validation and clearer error messages; robust handling of empty/edge cases and thread-count checks; new genomics, BED, VCF utilities and stats helpers.

* **Tests**
  * Extensive unit tests across the MSI pipeline, streaming intersection, utilities, and edge cases.

* **Chores**
  * Ignore local dev_sandbox/; added precise-decimal dependency for percentage calculations.

### CODE REVIEW

File: src/estimation/microsatellite_instability/intersection.rs [hunk 1]
```diff
@@ -0,0 +1,952 @@
...
+    for (i, &base) in changed_seq.iter().enumerate() {
+        let expected_base = motif_bytes[i % motif_len];
+        if base.to_ascii_uppercase() != expected_base {
+            // Not a perfect repeat: position {} has {} but expected {}
+            return RepeatStatus::NA;
+        }
+    }
+
+    RepeatStatus::Perfect
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:185]: It exists, review not applicable, as following is available: 
pub const fn [eq_ignore_ascii_case](https://doc.rust-lang.org/std/primitive.u8.html#method.eq_ignore_ascii_case)(&self, other: &[u8](https://doc.rust-lang.org/std/primitive.u8.html)) -> [bool](https://doc.rust-lang.org/std/primitive.bool.html)

File: src/estimation/microsatellite_instability/dp_analysis.rs [hunk 2]
```diff
@@ -0,0 +1,725 @@
...
+            .min(hundred)
+            .to_f64()
+            .unwrap_or(0.0);
+
+        let std_dev_f64 = std_dev_decimal.to_f64().unwrap_or(0.0);
+
+        (Some(lower), Some(upper), Some(std_dev_f64))
+    } else {
+        (None, None, None)
+    };
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/dp_analysis.rs]: I think only this is reasonable choice here: 
Explicitly guard this branch with an if total_regions > 0 check.

File: src/cli.rs [hunk 3]
```diff
@@ -528,6 +577,92 @@ pub enum EstimateKind {
...
+        data_distribution: Option<PathBuf>,
+        #[structopt(
+            long,
+            parse(from_os_str),
+            help = "Output path for pseudotime data (TSV), showing MSI \
+                    score evolution across allele frequency thresholds. \
+                    Recommended: .tsv extension."
+        )]
+        data_pseudotime: Option<PathBuf>,
+    },
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/cli.rs:689]: For comment 1, my program builds and runs for threshold constant. So, this comment is conservative.

[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/cli.rs:689]: Yes STDIN THING should be fixed.

As for other part, 
1. works 
--msi-threshold <msi-threshold>
            MSI classification threshold (must be > 0.0). [default: 3.5]
2. I think works as it is producing results as desired.
3. yes it succeeds

File: src/utils/bcf_utils.rs [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+            field: "DERIVED PROBABILITY ABSENT: PROB_ABSENT + PROBABILITY_ARTIFACT".to_string(),
+            value: probability as f32,
+            chrom: get_chrom(record, header)?,
+            pos: record.pos(),
+        }
+        .into());
+    }
+
+    Ok(Some(probability))
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs:240]: Fixed in [9399fa1](https://github.com/varlociraptor/varlociraptor/pull/527/commits/9399fa1e1af75bb02a7168a1967c327a16b2de23)

File: src/estimation/microsatellite_instability/intersection.rs [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+            &tmp_vcf.path().to_path_buf(),
+            &samples_index_map,
+        )
+        .unwrap();
+
+        // Verify no overlap
+        assert_eq!(total_regions, 1);
+        assert_eq!(results.len(), 0);
+    }
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs]: Fixed in [736df74](https://github.com/varlociraptor/varlociraptor/pull/527/commits/736df742348e058908f88c26362e454f3c65e0a5)

File: src/estimation/microsatellite_instability/intersection.rs [hunk 6]
```diff
@@ -0,0 +1,952 @@
...
+                        }
+                        None => {
+                            // Filtered
+                        }
+                    }
+                }
+            }
+        }
+        /* ------------------------------------------------------------- */
+        /* ------------------------------------------------------------- */
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:524]: fixed in [ffb227e](https://github.com/varlociraptor/varlociraptor/pull/527/commits/ffb227e17050e9a67a80b1589ab8808db334ed05)

File: src/estimation/microsatellite_instability/intersection.rs [hunk 7]
```diff
@@ -0,0 +1,952 @@
...
+                let pos = record.pos() as u64;
+                if pos < region.start {
+                    variant_window.pop_front();
+                } else {
+                    break;
+                }
+            } else {
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:425]: fixed in [ffb227e](https://github.com/varlociraptor/varlociraptor/pull/527/commits/ffb227e17050e9a67a80b1589ab8808db334ed05)

File: src/estimation/microsatellite_instability/dp_analysis.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/dp_analysis.rs]: fixed in [59a0fc4](https://github.com/varlociraptor/varlociraptor/pull/527/commits/59a0fc40bd3cae9c3537c613ac2050d176c57a6c)

File: src/estimation/microsatellite_instability/dp_analysis.rs [hunk 8]
```diff
@@ -0,0 +1,725 @@
...
+///
+/// Represents the probability that a microsatellite region is stable,
+/// calculated as the product of all variant absence probabilities.
+/// 
+/// p_stable = P(all variants absent) = Π(prob_absent)
+/// p_unstable = 1 - p_stable (computed in DP algorithm)
+#[derive(Debug, Clone)]
+struct RegionProbability {
+    p_stable: f64,
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/dp_analysis.rs:39]: fixed in [13d7996](https://github.com/varlociraptor/varlociraptor/pull/527/commits/13d7996366c1603c4527548b0954a798a4943955)

File: src/cli.rs [hunk 9]
```diff
@@ -43,6 +47,51 @@ use crate::variants::sample::{estimate_alignment_properties, MethylationReadtype
...
+}
+
+fn validate_thread_count(threads: Option<usize>) -> Result<()> {
+    if let Some(count) = threads {
+        if count < MIN_THREAD_COUNT {
+            bail!(errors::Error::InvalidThreadCount { count: (count) });
+        }
+    }
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/cli.rs:114]: fixed in [183144a](https://github.com/varlociraptor/varlociraptor/pull/527/commits/183144a1114121baf3d87c26d3e70bce779ee181)

File: src/cli.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/cli.rs:689]: fixed in [183144a](https://github.com/varlociraptor/varlociraptor/pull/527/commits/183144a1114121baf3d87c26d3e70bce779ee181)

File: src/estimation/microsatellite_instability/intersection.rs [hunk 10]
```diff
@@ -0,0 +1,974 @@
...
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:425]: Thank you for the review. I've verified this code compiles successfully with rust and passes tests without warnings.
The pattern you flagged does compile because:

Front loop (lines 383-407): The while let Some((record, chrom)) = variant_window.front() pattern creates an immutable borrow that is lexically scoped to each conditional branch. 

Back loop (lines 419-454): This also compiles.

File: templates/plots/msi_pseudotime.json [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                "dy": -5,
+                "fontSize": 11,
+                "color": "#d62728"
+            },
+            "encoding": {
+                "y": {
+                    "datum": 3.5
+                },
+                "text": { "value": "MSI Threshold" }
+            }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply templates/plots/msi_pseudotime.json:134]: Thank you for catching this. However, the `3.5` values in the templates are **intentional placeholders** that are dynamically replaced at runtime by the `write_plot` function.

File: templates/plots/msi_distribution.json [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                "align": "left",
+                "dx": 5,
+                "dy": -10,
+                "fontSize": 11,
+                "color": "#d62728"
+            },
+            "encoding": {
+                "x": { "datum": 3.5 },
+                "text": { "value": "MSI Threshold" }
+            }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply templates/plots/msi_distribution.json:65]: Thank you for catching this. However, the `3.5` values in the templates are **intentional placeholders** that are dynamically replaced at runtime by the `write_plot` function.

File: src/estimation/microsatellite_instability/intersection.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:425]: right!
yes
do!

File: templates/plots/msi_distribution.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply templates/plots/msi_distribution.json:65]: can you, yourself, resolve!

File: src/estimation/microsatellite_instability/intersection.rs [hunk 13]
```diff
@@ -0,0 +1,975 @@
...
+        return Ok(None);
+    }
+
+    let svlen = get_svlen(record, alt_idx, ref_allele, alt_allele)?;
+    let repeat_status = is_perfect_repeat(alt_allele, svlen, &region.motif, ref_allele);
+
+    if repeat_status == RepeatStatus::Perfect {
+        let prob_absent = get_prob_absent(record, header, alt_idx, is_phred)?;
+        let sample_afs = get_sample_afs(record, header, samples_index_map, alt_idx)?;
+
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs]: Ran in the new commit!

File: src/utils/genomics.rs [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+//     match normalized.parse::<u32>() {
+//         Ok(n) if (1..=22).contains(&n) => Some(n),
+//         _ => match normalized.as_str() {
+//             "X" | "x" => Some(23),
+//             "Y" | "y" => Some(24),
+//             "M" | "m" | "MT" | "Mt" | "mT" | "mt" => Some(25),
+//             _ => None,
+//         },
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/genomics.rs]: fixes in https://github.com/varlociraptor/varlociraptor/pull/527/commits/9249aa932aec07743829d56b3d02c7a29b47b21b

File: src/utils/genomics.rs [hunk 15]
```diff
@@ -0,0 +1,131 @@
+//! genomics.rs
+//!
+//! Genomics utility functions.
+//!
+//! This module provides utilities for:
+//! 1. Sequence analysis(Svlen calculation);
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/genomics.rs]: fixed in https://github.com/varlociraptor/varlociraptor/pull/527/commits/84f2eea503efd25076a1d60dc3481f29cb2cca05

File: src/errors.rs [hunk 16]
```diff
@@ -1,10 +1,131 @@
...
+    #[error(
+        "invalid VCF/BCF file path (expected .vcf, .vcf.gz, .bcf, or .bcf.gz extension): {path}"
+    )]
+    VcfFileInvalid { path: PathBuf },
+    #[error("VCF/BCF file contains no samples")]
+    VcfSamplesMissing,
+    #[error("VCF/BCF file is empty (no variant records)")]
+    VcfFileEmpty,
+    #[error("VCF/BCF record at {chrom}:{pos} is missing chromosome information")]
+    VcfRecordChromMissing { chrom: String, pos: i64 },
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/errors.rs]: fixed in https://github.com/varlociraptor/varlociraptor/pull/527/commits/9f0d78587bd399dcffde7521fd067cd7496595a6

File: src/utils/bcf_utils.rs [hunk 17]
```diff
@@ -0,0 +1,1034 @@
...
+        .map_err(|_| Error::VcfRecordChromResolveFailed {
+            pos: record.pos() + 1,
+            rid,
+            details: "Failed to resolve chromosome name".to_string(),
+        })?;
+
+    let chrom = String::from_utf8_lossy(chrom_bytes).to_string();
+
+    Ok(chrom)
+}
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs:96]: Check again I ran cargo fmt on bcf_utils

[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs:96]: The CI is passing format check, can you recheck !

File: src/utils/bcf_utils.rs [hunk 18]
```diff
@@ -0,0 +1,1034 @@
...
+        return Err(Error::VcfProbabilityValueInvalid {
+            field: "DERIVED PROBABILITY ABSENT: PROB_ABSENT + PROBABILITY_ARTIFACT".to_string(),
+            value: probability as f32,
+            chrom: get_chrom(record, header)?,
+            pos: record.pos() + 1,
+        }
+        .into());
+    }
+
+    let probability = probability.max(0.0).min(1.0);
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs]: fixed in [8c06ba2](https://github.com/varlociraptor/varlociraptor/pull/527/commits/8c06ba24cb26b496e7cd0e77358cb41b03d0659c)

File: src/utils/bcf_utils.rs [hunk 19]
```diff
@@ -0,0 +1,1034 @@
...
+
+    let probability_artifact = if is_phred {
+        phred_to_prob(prob_artifact as f64)
+    } else {
+        prob_artifact as f64
+    };
+
+    let probability = probability_absent + probability_artifact;
+
+    if probability < -EPSILON || probability > 1.0 + EPSILON {
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs]: fixed in [8c06ba2](https://github.com/varlociraptor/varlociraptor/pull/527/commits/8c06ba24cb26b496e7cd0e77358cb41b03d0659c)

File: src/estimation/microsatellite_instability/intersection.rs [hunk 20]
```diff
@@ -0,0 +1,953 @@
...
+/// * `RepeatStatus::Perfect` - Changed sequence is exact motif repeats
+/// * `RepeatStatus::NA` - Does not match or fails validation
+///
+/// # Example
+/// assert_eq!(is_perfect_repeat(b"ACAGCAG", 3, "CAG", b"ACAG"), RepeatStatus::Perfect);
+fn is_perfect_repeat(alt_seq: &[u8], svlen: i32, motif: &str, ref_seq: &[u8]) -> RepeatStatus {
+    // 0. Handling Edge Cases
+    if ref_seq.is_empty() || alt_seq.is_empty() || ref_seq[0] != alt_seq[0] || svlen == 0 {
+        return RepeatStatus::NA;
+    }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/intersection.rs:134]: fixed in [4d24adf](https://github.com/varlociraptor/varlociraptor/pull/527/commits/4d24adf8743f6db9163c0dd858c4e14c35cc150e)

File: src/utils/bcf_utils.rs [hunk 21]
```diff
@@ -0,0 +1,1035 @@
...
+
+    let valid_range = -EPSILON..=1.0 + EPSILON;
+    if !valid_range.contains(&probability) {
+        return Err(Error::VcfProbabilityValueInvalid {
+            field: "DERIVED PROBABILITY ABSENT: PROB_ABSENT + PROBABILITY_ARTIFACT".to_string(),
+            value: probability as f32,
+            chrom: get_chrom(record, header)?,
+            pos: record.pos() + 1,
+        }
+        .into());
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/utils/bcf_utils.rs:234]: fixed in [2f58a33](https://github.com/varlociraptor/varlociraptor/pull/527/commits/2f58a33564ed6236a1ff66d7425f4a2c233f125d)

File: src/estimation/microsatellite_instability/output.rs [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+                sample,
+                af_f64,
+                result.msi_score_map.unwrap(),
+                result.k_map.unwrap(),
+                result.regions_with_variants.unwrap(),
+                result.msi_status.as_ref().unwrap(),
+                lower,
+                upper,
+                std_dev
+            )?;
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:377]: In our case, it is only called when config.data_pseudotime.is_some(), hence isn't it already safe and defensive?

File: src/estimation/microsatellite_instability/output.rs [hunk 23]
```diff
@@ -0,0 +1,472 @@
...
+
+            let msi_score = result.msi_score_map.unwrap();
+            data.push(json!({
+                "sample": sample,
+                "af_threshold": af_f64,
+                "msi_score": result.msi_score_map,
+                "lower_bound": result.uncertainty_lower.unwrap_or(msi_score),
+                "upper_bound": result.uncertainty_upper.unwrap_or(msi_score),
+            }));
+        }
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:456]: In our case, it is only called when config.data_pseudotime.is_some(), hence isn't it already safe and defensive?

File: src/estimation/microsatellite_instability/output.rs [hunk 24]
```diff
@@ -0,0 +1,472 @@
...
+        writer,
+        "sample\tk\tmsi_score(threshold={:.1})\tprobability",
+        msi_threshold
+    )?;
+
+    for (sample, af_results) in results {
+        let af_zero_result = match af_results.get("0") {
+            Some(result) => result,
+            None => continue,
+        };
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:185]: fixed in [fd7373b](https://github.com/varlociraptor/varlociraptor/pull/527/commits/fd7373badfee29103fc17a228c1c561ba0ea678a)

File: src/estimation/microsatellite_instability/output.rs [hunk 25]
```diff
@@ -0,0 +1,472 @@
...
+    path: &Path,
+    msi_threshold: f64,
+) -> Result<()> {
+    let mut data = Vec::new();
+
+    for (sample, af_results) in results {
+        let af_zero_result = match af_results.get("0") {
+            Some(result) => result,
+            None => continue,
+        };
```
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:248]: fixed in [fd7373b](https://github.com/varlociraptor/varlociraptor/pull/527/commits/fd7373badfee29103fc17a228c1c561ba0ea678a)

File: src/estimation/microsatellite_instability/output.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:456]: so are you closing this!

File: src/estimation/microsatellite_instability/output.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (rohan-ibn-tariq) [reply src/estimation/microsatellite_instability/output.rs:377]: are you resolving this?
